src/transaction/script.c

```diff
@@ -0,0 +1,228 @@
+    if ((in == NULL) || (out == NULL) || (script_len < expected_len) || (script_len > MAX_DATA_SCRIPT_LEN) || (in->size - in->offset < script_len)) {
+        // Maybe return error code?
+        return 0;
+    }
+    if (in->ptr[in->offset] == OP_PUSHDATA1) {
+      // If first opcode is OP_PUSHDATA1, the second is the data length
+      // then we push the actual data
+      // Valid for lengths [1,255] but used in (75,255] due to extra opcode byte
+      // expected_len is increased due to the extra OP_PUSHDATA1
+      expected_len += 1;
+      data_len = in->ptr[in->offset + 1];
+      data_start = 2;
+    } else {
+      // The first opcode is the data length
+      // then we push the actual data
+      // This case is only valid for [1,75] bytes of data
+      data_len = in->ptr[in->offset];
+      data_start = 1;
+      if (data_len > 75) {
+          // Invalid script
+          // Maybe return error code?
+          return 0;
+      }
+    }
+    expected_len += data_len;
+    if (script_len != expected_len) {
+        // Invalid script
+        // Here we check for actual data push and data + OP_CHECKSIG length
+        return 0;
+    }
+
+    // Read actual data from buffer
+    memmove(out, in->ptr + in->offset + data_start, data_len);
+    return data_len;
+}
+
+/**
+ * Identifies the given script type
+ */
+script_type_t identify_script(buffer_t *in, size_t script_len) {
+    if (identify_p2pkh_script(in, script_len)) {
+        return SCRIPT_P2PKH;
+    } else if (identify_p2sh_script(in, script_len)) {
+        return SCRIPT_P2SH;
+    } else if (identify_data_script(in, script_len)) {
+        return SCRIPT_DATA;
+    } else {
+        // Unsupported script should fail output parsing
+        return SCRIPT_UNKNOWN;
+    }
+}
+
+uint16_t parse_output_script(buffer_t *in, size_t script_len, output_script_info_t *out) {
+    if (in == NULL || out == NULL) {
+        // Choose an error code for invalid arguments
+        return ERR_INVALID_ARGS;
+    }
+
+    switch(identify_script(in, script_len)) {
+      case SCRIPT_P2PKH:
+          out->type = SCRIPT_P2PKH;
+          if (script_len != 25) {
+              // timelock is not supported yet
+              return ERR_UNSUPPORTED_SCRIPT;
+          }
+          memmove(out->hash, in->ptr + in->offset + 3, PUBKEY_HASH_LEN);
+          break;
+      case SCRIPT_P2SH:
+          out->type = SCRIPT_P2SH;
+          if (script_len != 23) {
+              // timelock is not supported yet
+              return ERR_UNSUPPORTED_SCRIPT;
+          }
+          memmove(out->hash, in->ptr + in->offset + 2, PUBKEY_HASH_LEN);
+          break;
+      case SCRIPT_DATA:
+          out->type = SCRIPT_DATA;
+          G_create_token_data.data_len = read_data_script(
+              in,
+              script_len,
+              G_create_token_data.data);
+          if (G_create_token_data.data_len == 0) {
+              return ERR_UNSUPPORTED_SCRIPT;
+          }
+          break;
```

Comment:
We read data outputs into the global context because we only need support for 1 data output.
If this changes we will need to change this part.

Having a `MAX_DATA_SCRIPT_LEN` (153 bytes) empty array on all outputs will make the RAM usage way higher than needed.
This is why we use the global context instead of the output in the data script case.